### DISCUSSION TITLE
Big Sur brew commands try to link arm64e, even though on x86_64 machine

### DISCUSSION THREAD

**Discussion by mbarrien** [Homebrew/discussions#932] [Everyday usage] [answered]
I've run into a problem running Homebrew on a x86_64 based MacBook Pro, where running most brew commands results in it failing to install Ruby gem packages that have native extensions. This leaves Homebrew unable to even run "brew --version" or "brew doctor".

The underlying error message indicates that the linker is trying to link binary files built for arm64e architecture to underlying libraries that were built for x86_64. This is strange since the machine I'm using for this has always been x86_64, and had a working brew command line until I recently tried to run brew update.

The key error message, repeated for several different gems, is:
```
ld: warning: ignoring file cparse.o, building for macOS-arm64e but attempting to link with file built for unknown-x86_64
ld: in '/usr/local/lib/libunwind.dylib', building for macOS-arm64e but attempting to link with file built for macOS-x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

A full gist of a terminal log (including some diagnostic commands, and proof that I cleaned up old cached gems and have updated all repos) can be found at <https://gist.github.com/mbarrien/5d8cb2aeda679017d411a4af8f594c4e>.

I narrowed down the issue to the Makefile generated for these gems. It contains the line
```
dldflags = -arch x86_64 -arch arm64e             -undefined dynamic_lookup -multiply_defined suppress
```

If I remove the  `-arch arm64e` and leave the `-arch x86_64` and manually run make, this build with no errors. If I leave it in, I get the error above.

```
➜  fast_xs git:(master) make
linking shared-object fast_xs.bundle
ld: warning: directory not found for option '-L/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.Internal.sdk/usr/local/lib'
ld: warning: directory not found for option '-L/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.Internal.sdk/usr/local/lib'
ld: warning: ignoring file fast_xs.o, building for macOS-arm64e but attempting to link with file built for unknown-x86_64
ld: in '/usr/local/lib/libunwind.dylib', building for macOS-arm64e but attempting to link with file built for macOS-x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [fast_xs.bundle] Error 1
➜  fast_xs git:(master) sed -i '' -e 's/ -arch arm64e//' Makefile
➜  fast_xs git:(master) make
linking shared-object fast_xs.bundle
ld: warning: directory not found for option '-L/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.Internal.sdk/usr/local/lib'
➜  fast_xs git:(master)
```

It's almost as if the linker only looks at the last -arch instead of considering them both for a universal build.

**Comment by Bo98** [comment, marked as answer]
This is a macOS Ruby bug - FB9002047. I'm surprised it's causing fatal errors though. Usually the only probelem is it using `arm64e` when it should be `arm64`.

Try `brew update` now though - we've disabled gem installations for most use cases.

**Reply by mbarrien** [reply]
Normal `brew update` triggered the same error. However, after running `brew update-reset` first, `brew update` started working, and it now works on my machine.

Thanks!
